### PR TITLE
#2516 target android 11

### LINKED ISSUE #2516: Target Android 11 before November 2021
**Is your feature request related to a problem? Please describe.**
To continue to update kiwix we will have to target Android 11 and consequently SAF. https://android-developers.googleblog.com/2020/11/new-android-app-bundle-and-target-api.html

### PR DESCRIPTION
**Pull request by gouri-panda**
Fixes #2516

### CODE REVIEW

**Comment by gouri-panda** [comment]
@kelson42 In this patch we have to think about #2646 . More info found [here](https://developer.android.com/about/versions/11/privacy/storage).

**Comment by kelson42** [comment]
@gouri-panda I DON't really underatand this documentation. What does this patch change for the Kiwix end-user? What does it change for the devs?

**Comment by gouri-panda** [comment]
@kelson42 In Android 11 android updated its storage permission. Starting in Android 11, If the app needs to manage device storage, we have to add `MANAGE_EXTERNAL_STORAGE` permission. Which we discussed in that other tickets. We have to add that permission if we wanna access external storage permission, and we have to tell the play store why we need this permission. More info [here](https://developer.android.com/about/versions/11/privacy/storage). 

 Now in android 11 starting from November Google Play will require new apps to use the Android App Bundle(a.k.a aab) publishing format. In that case, we will reduce 15% average application size according to google. Which will big benefit for `kiwix-android` and `kiwix-custom`  More info [here](https://android-developers.googleblog.com/2020/11/new-android-app-bundle-and-target-api.html).

**Comment by kelson42** [comment]
> @kelson42 In Android 11 android updated its storage permission. Starting in Android 11, If the app needs to manage device storage, we have to add `MANAGE_EXTERNAL_STORAGE` permission. Which we discussed in that other tickets. We have to add that permission if we wanna access external storage permission, and we have to tell the play store why we need this permission. More info [here](https://developer.android.com/about/versions/11/privacy/storage).

OK, what should we do now? Can we not just add this permission for the moment?
 
> Now in android 11 starting from November Google Play will require new apps to use the Android App Bundle(a.k.a aab) publishing format. In that case, we will reduce 15% average application size according to google. Which will big benefit for `kiwix-android` and `kiwix-custom` More info [here](https://android-developers.googleblog.com/2020/11/new-android-app-bundle-and-target-api.html).

I guess a new ticket should be created for that? Does not sound trivial to do.

**Comment by gouri-panda** [comment]
>OK, what should we do now? Can we not just add this permission for the moment?
yes
>I guess a new ticket should be created for that? Does not sound trivial to do.
Sean discussed in that ticket. So I thought it's the right moment to do. Now If you think we should create a new ticket for this. I'm totally fine :)

**Comment by kelson42** [comment]
> > I guess a new ticket should be created for that? Does not sound trivial to do.
> 
> Sean discussed in that ticket. So I thought it's the right moment to do. Now If you think we should create a new ticket for this. I'm totally fine :)

You mean #2516? I see no comment or information from Sean. If this is quick&easy then I'm in favour to do it now. Otherwise no. We have currently a few serious regressions and focus should be on the GSoC topic.

**Comment by gouri-panda** [comment]
>You mean #2516? I see no comment or information from Sean.

In the first message. He pointed out this.
>We have currently a few serious regressions and focus should be on the GSoC topic.

Ok

**Comment by gouri-panda** [comment]
I will resume this work after August.

Note to myself:
Create a separate ticket  for the Android App Bundle( Playstore upload format)

**Comment by kelson42** [comment]
@gouri-panda Sorry missed your last commnent. If possible (not to much work), would be great to do it now... so we can close https://github.com/kiwix/kiwix-android/milestone/12

**Comment by kelson42** [comment]
@gouri-panda What is the status here? I see no reviewer assigned and I'm not even sure it fully fix #2516?!

**Comment by gouri-panda** [comment]
>What is the status here? 

@kelson42  I was waiting for #2668 .

> I see no reviewer assigned

Done 👍 
> I'm not even sure it fully fix #2516?!

 What did I miss?

**Comment by kelson42** [comment]
@gouri-panda Thx, great if it just fixes #2516

**Comment by kelson42** [comment]
@mhutti1 @abdulwd Can we please have a review?

File: app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/KiwixWifiP2pBroadcastReceiver.kt [hunk 1]
```diff
@@ -53,7 +53,9 @@ class KiwixWifiP2pBroadcastReceiver(private val p2pEventListener: P2pEventListen
       WIFI_P2P_CONNECTION_CHANGED_ACTION -> {
         val networkInfo =
           intent.getParcelableExtra<NetworkInfo>(EXTRA_NETWORK_INFO)
-        p2pEventListener.onConnectionChanged(networkInfo.isConnected)
+        if (networkInfo != null) {
```
Comment (abdulwd) [inline app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/KiwixWifiP2pBroadcastReceiver.kt:56]: let can be used to make it Kotliny

File: app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/ReceiverDevice.kt [hunk 2]
```diff
@@ -72,7 +72,9 @@ internal class ReceiverDevice(private val wifiDirectManager: WifiDirectManager)
                   publishProgress(fileItemIndex, FileItem.FileStatus.SENT)
                 }
               } catch (e: IOException) {
-                Log.e(TAG, e.message)
+                e.message?.let { message ->
+                  Log.e(TAG, message)
```
Comment (abdulwd) [inline app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/ReceiverDevice.kt:76]: A string identifier can be added to this message some like "Unable to do something: $message" also pass exception as the 3rd param to `Log.e`

File: app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/ReceiverDevice.kt [hunk 3]
```diff
@@ -81,7 +83,9 @@ internal class ReceiverDevice(private val wifiDirectManager: WifiDirectManager)
         }
       }
     } catch (e: IOException) {
-      Log.e(TAG, e.message)
+      e.message?.let { message ->
+        Log.e(TAG, message)
```
Comment (abdulwd) [inline app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/ReceiverDevice.kt:87]: Same for here

File: core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixTextToSpeech.kt [hunk 4]
```diff
@@ -84,7 +84,7 @@ class KiwixTextToSpeech internal constructor(
    * Reads the currently selected text in the WebView.
    */
   fun readSelection(webView: WebView) {
-    webView.loadUrl("javascript:tts.speakAloud(window.getSelection().toString());", null)
+    webView.loadUrl("javascript:tts.speakAloud(window.getSelection().toString());")
```
Comment (abdulwd) [inline core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixTextToSpeech.kt:87]: Why was this change required?

File: core/src/main/java/org/kiwix/kiwixmobile/core/search/viewmodel/effects/ProcessActivityResult.kt [hunk 5]
```diff
@@ -38,7 +38,10 @@ data class ProcessActivityResult(
       resultCode == Activity.RESULT_OK &&
       data != null
     ) {
-      actions.offer(Filter(data.getStringArrayListExtra(RecognizerIntent.EXTRA_RESULTS)[0]))
+      val results = data.getStringArrayListExtra(RecognizerIntent.EXTRA_RESULTS)
+      if (results != null && results.first() != null) {
+        actions.offer(Filter(results.first()))
+      }
```
Comment (abdulwd) [inline core/src/main/java/org/kiwix/kiwixmobile/core/search/viewmodel/effects/ProcessActivityResult.kt:44]: **Optional:** We can try to write this using let, take & other [scope functions in Kotlin](https://kotlinlang.org/docs/scope-functions.html)

**Review by abdulwd — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixTextToSpeech.kt (hunk 4 above)
Comment (gouri-panda) [reply core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixTextToSpeech.kt:87]: Android 11  handles null differently. It doesn't compile if we pass the null value in `additionalHttpHeaders`.
Since we don't need `additionalHttpHeaders` we can remove them
The below error shows if we pass null value
```
Null can not be a value of a non-null type (Mutable)Map<String!, String!>
```

Comment (gouri-panda) [reply core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixTextToSpeech.kt:87]: According to android
>If you are writing in Kotlin, when upgrading from the Android 10 to the Android 11 SDK, you may notice that there are some new compiler warnings and that previous warnings may have been upgraded to errors. This is intended and a feature of the Kotlin compiler — these warnings tell you that you may be writing code that crashes your app at runtime (a risk you would miss entirely if you weren’t writing in Kotlin). As you encounter these warnings and errors, you can handle them by adding null checks to your code.

**Comment by kelson42** [comment]
@mhutti1 @abdulwd Any problem to get anew review? This has been requested 5 days ago already!

File: app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/ReceiverDevice.kt [hunk 6]
```diff
@@ -72,7 +72,9 @@ internal class ReceiverDevice(private val wifiDirectManager: WifiDirectManager)
                   publishProgress(fileItemIndex, FileItem.FileStatus.SENT)
                 }
               } catch (e: IOException) {
-                Log.e(TAG, e.message)
+                e.message?.let { message ->
+                  Log.e(TAG, "I/O exception $message", e)
+                }
```
Comment (abdulwd) [inline app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/ReceiverDevice.kt:77]:
```suggestion
                  Log.e(TAG, "<Unable to accomplish the operation we were trying to do>", e)
```

**Review by abdulwd — CHANGES_REQUESTED** [review]
@gouri-panda can you please quickly pass in a relevant log message to the logger wherever you're using **Log.**

Looks good otherwise. 👍

File: .github/workflows/release.yml [hunk 7]
```diff
@@ -36,14 +36,14 @@ jobs:
           KEY_ALIAS: ${{ secrets.KEY_ALIAS }}
           KEY_PASSWORD: ${{ secrets.KEY_PASSWORD }}
           KEY_STORE_PASSWORD: ${{ secrets.KEY_STORE_PASSWORD }}
-        run: ./gradlew publishReleaseApk
+        run: ./gradlew publishReleaseBundle
 
       - name: Publish app to download.kiwix.org
         env:
-          UNIVERSAL_RELEASE_APK: app/build/outputs/apk/release/*universal*.apk
+          UNIVERSAL_RELEASE_AAB: app/build/outputs/apk/release/**.aab
```
Comment (s-ayush2903) [inline .github/workflows/release.yml:43]: I think it is never going to work, as the path being provided here to locate application bundle is incorrect, as per the [_official_ article](https://medium.com/androiddevelopers/building-your-first-app-bundle-bbcd228bf631)

See:
image

**Comment by s-ayush2903** [comment]
I believe it'll be better if we simulate a sample aab publish to GitHub CI as artifacts, in order to avoid last minute crunch when we're creating a release and current release config doesn't work for some reasons then :P
(No offfense intended, just mentioning to make sure we have a 100% working release config)

**Comment by gouri-panda** [comment]
Thanks @s-ayush2903 will do!

File: .github/workflows/release.yml (hunk 7 above)
Comment (gouri-panda) [reply .github/workflows/release.yml:43]: Thanks!

**Review by s-ayush2903 — COMMENTED** [review]
You'll need to change GitHub Release job as well, I don't know why Github doesn't let me comment on that section of code, so I'm attaching the picture here for reference and your ease.
See:
image

----

Also, it'll be better if you try uploading the aab in this PR via actions to GH CI as artifacts. Because I'm not _very_ sure how this aab thing is going to work, as previously when we published APKs to play store, then it was our `universal` build, which you removed here(in line#43 of the same release workflow file) and changed to the generic aab, so it'll be better to upload the entire aab and then explore its components and then upload the relevant ones to play store.
Let me know if you need any help creating a new job for simulating the process in this PR :v:

**Comment by gouri-panda** [comment]
>You'll need to change GitHub Release job as well, I don't know why Github doesn't let me comment on that section of code, so I'm attaching the picture here for reference and your ease.

Thanks!
>Also, it'll be better if you try uploading the aab in this PR via actions to GH CI as artifacts. Because I'm not very sure how this aab thing is going to work, as previously when we published APKs to play store, then it was our universal build, which you removed here(in line#43 of the same release workflow file) and changed to the generic aab, so it'll be better to upload the entire aab and then explore its components and then upload the relevant ones to play store.

I'm not sure I can't make this decision. Perhaps @kelson42 Help us out here! I think google play store handles that for us.

**Comment by kelson42** [comment]
What is the aab file?

**Comment by gouri-panda** [comment]
>What is the aab file?

Android app bundle. It's Android's new App Publish Format-like we do now in apk.[see](https://developer.android.com/platform/technology/app-bundle)

**Comment by kelson42** [comment]
Ok, so does the abb file replaces the apk or should it be done in addition?

**Comment by gouri-panda** [comment]
>so does the abb file replaces the apk 
Yes

**Comment by kelson42** [comment]
If it works the same, we should do the same as today:
* upload to download.kiwix.org (nightly, multi-arch release)
* upload to play store if requested
* make proper releases on github

**Comment by gouri-panda** [comment]
According to the [official](https://developer.android.com/guide/app-bundle) doc
>Google Play uses your app bundle to generate and serve optimized APKs for each device configuration, so only the code and resources that are needed for a specific device are downloaded to run your app. You no longer have to build, sign, and manage multiple APKs to optimize support for different devices, and users get smaller, more-optimized downloads.

@s-ayush2903 I think we don't need Universal APK to manage multiple APK anymore.

**Comment by s-ayush2903** [comment]
Okay, looks fine to me. Just to avoid that grey area and to be hundred percent sure, can we please see an `.aab` being uploaded as a CI artifact?

**Comment by kelson42** [comment]
> I think we don't need Universal APK to manage multiple APK anymore.

@gouri-panda What is the status of this PR? We need a universal AAB to be able to distribute outside the Google Play store.

**Comment by gouri-panda** [comment]
>We need a universal AAB to be able to distribute outside the Google Play store.

@kelson42 No. We don't need anymore. Google Play store handles that for us.

**Comment by kelson42** [comment]
@gouri-panda This is a requirement: we need a way to install Kiwix outside the Play store. So, how to secure this?

**Comment by gouri-panda** [comment]
>@gouri-panda This is a requirement: we need a way to install Kiwix outside/without the Play store. So, how to secure this?


I don't know the other app store's application format. But Amazon app store will [support](https://developer.amazon.com/blogs/appstore/post/50b1ca0f-bbec-48ec-9eea-3c395efb8f9f/amazon-appstore-to-support-android-app-bundle) AAB. Since it's now the official format for the Google play store the other app store will have to support this format. Amazon is saying that they both support APK & AAB. So I guess other app stores such as F-Droid , Huawei &  Samsung Galaxy Store will have to support this way. @abdulwd @mhutti1 Can you give some suggestions about this? Thanks!

**Comment by kelson42** [comment]
> > @gouri-panda This is a requirement: we need a way to install Kiwix outside/without the Play store. So, how to secure this?
> 
> I don't know the other app store's application format. But Amazon app store will [support](https://developer.amazon.com/blogs/appstore/post/50b1ca0f-bbec-48ec-9eea-3c395efb8f9f/amazon-appstore-to-support-android-app-bundle) AAB. Since it's now the official format for the Google play store the other app store will have to support this format. Amazon is saying that they both support APK & AAB. So I guess other app stores such as F-Droid , Huawei & Samsung Galaxy Store will have to support this way. @abdulwd @mhutti1 Can you give some suggestions about this? Thanks!

This is not about an other store, this is about being able to install Kiwix for Android *without any* store.

**Comment by gouri-panda** [comment]
>This is not about an other store, this is about being able to install Kiwix for Android without any store.

 We can download AAB from download.kiwix.org. But unfortunately, we can't install directly without the google play store.
We have to follow this [procedure](https://stackoverflow.com/a/50465619/10567024) to install the app on device.

**Comment by kelson42** [comment]
> > This is not about an other store, this is about being able to install Kiwix for Android without any store.
> 
> We can download AAB from download.kiwix.org. But unfortunately, we can't install directly without the google play store.
> We have to follow this [procedure](https://stackoverflow.com/a/50465619/10567024) to install the app on device.

OK, then we need to continue to build and publish APKs on download.kiwix.org. Should not be a big deal.

**Comment by gouri-panda** [comment]
@kelson42 Now in our Google play store format is AAB and we store it in `download.kiwix.org` in APK format(Universal APK).

File: .github/workflows/release.yml [hunk 8]
```diff
@@ -36,7 +36,7 @@ jobs:
           KEY_ALIAS: ${{ secrets.KEY_ALIAS }}
           KEY_PASSWORD: ${{ secrets.KEY_PASSWORD }}
           KEY_STORE_PASSWORD: ${{ secrets.KEY_STORE_PASSWORD }}
-        run: ./gradlew publishReleaseApk
+        run: ./gradlew publishReleaseBundle ; ./gradlew publishReleaseApk
```
Comment (s-ayush2903) [inline .github/workflows/release.yml:39]: Also can you tell which out of these two is going to be published at playstore? I think we should separate out these two commands, otherwise maybe it'll push the result of latst command? Again, to avoid grey area, I'll re-iterate about the simulation in the PR

File: .github/workflows/release.yml [hunk 9]
```diff
@@ -48,7 +48,7 @@ jobs:
       - name: Publish to github releases
         uses: ncipollo/release-action@v1
         with:
-          artifacts: "app/build/outputs/apk/release/**"
+          artifacts: "app/build/outputs/bundle/release/**"
```
Comment (s-ayush2903) [inline .github/workflows/release.yml:51]: As now we've decided to create a normal apk based multi-arc release, this change should be reverted.

File: .github/workflows/release.yml (hunk 8 above)
Comment (gouri-panda) [reply .github/workflows/release.yml:39]: Yes, we can separate those commands. This is making confusing.

File: .github/workflows/release.yml [hunk 10]
```diff
@@ -36,12 +36,16 @@ jobs:
           KEY_ALIAS: ${{ secrets.KEY_ALIAS }}
           KEY_PASSWORD: ${{ secrets.KEY_PASSWORD }}
           KEY_STORE_PASSWORD: ${{ secrets.KEY_STORE_PASSWORD }}
-        run: ./gradlew publishReleaseApk
+        run: ./gradlew publishReleaseBundle
 
       - name: Publish app to download.kiwix.org
         env:
+          KEY_ALIAS: ${{ secrets.KEY_ALIAS }}
```
Comment (kelson42) [inline .github/workflows/release.yml:43]: I don’t understand why we need this now. This ruke should not have changed?!

**Review by kelson42 — CHANGES_REQUESTED** [review]
@gouri-panda Please modify your Pr to test an apk upload of an apk to the nightlies… if it works, then you will be able to modify back so we can be ready for merge.

**Comment by kelson42** [comment]
@gouri-panda Can we move on relatively quickly now on this? It seems you are almost done and I would like to merge soon to be able to release 3.4.5.

**Comment by gouri-panda** [comment]
@kelson42 Will do! Sorry, about the delay, and thanks for the review!

File: .github/workflows/release.yml (hunk 10 above)
Comment (gouri-panda) [reply .github/workflows/release.yml:43]: I thought we need an alias name. Ok, I'll remove that 👍 .

**Comment by gouri-panda** [comment]
>Please modify your Pr to test an apk upload of an apk to the nightlies…

@kelson42 Is it possible in PR? If so can you tell me how? Thanks!

**Comment by kelson42** [comment]
@gouri-panda Just publish the APK on download.kiwix.org based on the PR commit event instead of on the nightly cron event, like @s-ayush2903 did in his latest PR.

**Comment by kelson42** [comment]
@gouri-panda where is the nightly apk you have published from the branch?

**Review by s-ayush2903 — CHANGES_REQUESTED** [review]
@gouri-panda You probably misinterpreted what @kelson42 said. 
You were supposed to publish the apk corresponding to the changes made in _this PR_, but what you did in eb7daf7 does nothing other than publishing apk for the changes that _already_ exist in develop. 
As mentioned previously in this PR a few times, kindly publish the latest config from _this_ PR to [download.kiwix.org](https://download.kiwix.org). Also, kindly clarify with the reviewer if there is anything unclear before making any major changes, nightly builds use a lot of resources and freeze all the other jobs, resulting in longer queues. 
* As an fyi, the latest commit you pushed now is going to overwrite the existing apk published at [download.kiwix.org](https://download.kiwix.org) with this one, without verifying any changes relevant to this PR.

**Comment by gouri-panda** [comment]
@s-ayush2903  sorry! Won't happen again.

**Comment by kelson42** [comment]
@gouri-panda any news here, we really need to move on?

**Comment by gouri-panda** [comment]
@kelson42 For some reason the permission denies. see [here](https://github.com/kiwix/kiwix-android/pull/2660/checks?check_run_id=3293816911)

**Comment by kelson42** [comment]
@gouri-panda no clue... if it work for the nightly it should work as well here from download.kiwix,org perspective

**Comment by kelson42** [comment]
@s-ayush2903 Would you be able to help @gouri-panda on this critical PR please?

**Review by kelson42 — APPROVED** [review]
